light-client/src/builder/light_client.rs

```diff
@@ -0,0 +1,213 @@
+        light_store: Box<dyn LightStore>,
+        io: Box<dyn Io>,
+        hasher: Box<dyn Hasher>,
+        clock: Box<dyn Clock>,
+        verifier: Box<dyn Verifier>,
+        scheduler: Box<dyn Scheduler>,
+        predicates: Box<dyn VerificationPredicates>,
+    ) -> Self {
+        Self {
+            peer_id,
+            hasher,
+            io,
+            verifier,
+            light_store,
+            clock,
+            scheduler,
+            options,
+            predicates,
+            state: NoTrustedState,
+        }
+    }
+
+    /// Set the given light block as the initial trusted state.
+    pub fn trust_light_block(
+        mut self,
+        trusted_state: LightBlock,
+    ) -> Result<LightClientBuilder<HasTrustedState>, Error> {
+        self.validate(&trusted_state)?;
+
+        // TODO(liamsi, romac): it is unclear if this should be Trusted or only Verified
+        self.light_store.insert(trusted_state, Status::Trusted);
+
+        Ok(self.with_state(HasTrustedState))
+    }
+
+    /// Keep using the latest verified or trusted block in the light store.
+    /// Such a block must exists otherwise this will fail.
+    pub fn trust_from_store(self) -> Result<LightClientBuilder<HasTrustedState>, Error> {
+        let trusted_state = self
+            .light_store
+            .latest_trusted_or_verified()
+            .ok_or_else(|| error::Kind::NoTrustedStateInStore)?;
+
+        self.trust_light_block(trusted_state)
+    }
+
+    /// Fetch and set the latest block from the primary peer as the trusted state.
+    pub fn trust_primary_latest(self) -> Result<LightClientBuilder<HasTrustedState>, Error> {
+        let trusted_state = self
+            .io
+            .fetch_light_block(AtHeight::Highest)
+            .map_err(error::Kind::Io)?;
+
+        self.trust_light_block(trusted_state)
+    }
+
+    /// Set the block from the primary peer at the given height as the trusted state.
+    pub fn trust_primary_at(
+        self,
+        trusted_height: Height,
+        trusted_hash: Hash,
+    ) -> Result<LightClientBuilder<HasTrustedState>, Error> {
+        let trusted_state = self
+            .io
+            .fetch_light_block(AtHeight::At(trusted_height))
+            .map_err(error::Kind::Io)?;
+
+        if trusted_state.height() != trusted_height {
+            bail!(error::Kind::HeightMismatch {
+                given: trusted_height,
+                found: trusted_state.height(),
+            });
+        }
+
+        let header_hash = self.hasher.hash_header(&trusted_state.signed_header.header);
+
+        if header_hash != trusted_hash {
+            bail!(error::Kind::HashMismatch {
+                given: trusted_hash,
+                found: header_hash,
+            });
+        }
+
+        self.trust_light_block(trusted_state)
+    }
+
+    fn validate(&self, light_block: &LightBlock) -> Result<(), Error> {
+        // TODO(ismail, romac): actually verify more predicates of light block?
```

Comment:
We should probably check:
- is_within_trust_period
- is_header_from_past
- validator_sets_match (though this doesn't really matter since we will never use these vals)
- next_validators_match

Everything else depends on the commit and/or previous blocks. We could verify the commit but it's not really necessary.

Reply:
3f496fd